runtime/participant-runtime.ss

```diff
@@ -767,3 +767,380 @@
+(def (send-contract-agreement agreement off-chain-channel)
+  (displayln MAGENTA "Sending Agreement to other participants...")
+  (match (.@ off-chain-channel tag)
+    ('stdio (send-contract-agreement/stdout agreement))
+    ('libp2p (let ()
+     (def libp2p-client (.@ off-chain-channel libp2p-client))
+     (def dest-address (.@ off-chain-channel dest-address))
+     (send-contract-agreement/libp2p agreement libp2p-client dest-address)))
+    (else (error "Invalid channel")))) ; TODO: This is an internal error,
+                                       ; ensure this is handled at cli options parsing step.
+
+(def (send-contract-agreement/stdout agreement)
+  (displayln MAGENTA "One line command for other participants to generate the same agreement:" END)
+  (def agreement-string (string<-json (json<- InteractionAgreement agreement)))
+  (def escaped-agreement-string (escape-json-string agreement-string))
+  (def full-cmd-string (string-append "glow start-interaction --agreement " escaped-agreement-string))
+  (displayln full-cmd-string)
+  (force-output))
+
+;; TODO: Refactor this to be more general, handshake uses this as well.
+(def (send-contract-agreement/libp2p agreement libp2p-client dest-address)
+  (displayln MAGENTA "Sending agreement to multiaddr..." END)
+  (def agreement-string (string<-json (json<- InteractionAgreement agreement)))
+  ;; (if (string-contains agreement-string "'")
+  ;;   (pr agreement-string)
+  ;;   (display (string-append "'" agreement-string "'")))
+  (dial-and-send-contents libp2p-client dest-address agreement-string)
+  (displayln)
+  (force-output))
+
+;; TODO: Generalize & Upstream to gerbil-utils/json
+(def (escape-json-string json-string)
+  (if (string-contains json-string "'")
+    json-string
+    (string-append "'" json-string "'")))
+
+
+;; ------------------ Listen for agreements
+
+
+(def (listen-for-agreement off-chain-channel)
+  (match (.@ off-chain-channel tag)
+    ('stdio
+     (let ()
+       (displayln MAGENTA "Listening for agreement via stdin ...")
+       (def agreement-json (parameterize ((json-symbolic-keys #f)) (read-json)))
+       (def agreement (<-json InteractionAgreement agreement-json))
+       agreement))
+    ('libp2p
+     (let ()
+       (displayln MAGENTA "Listening for agreement via libp2p ...")
+       (def agreement-str (.call off-chain-channel poll-buffer))
+       (displayln MAGENTA "Received agreement")
+       (def agreement (<-json InteractionAgreement (json<-string agreement-str)))
+       agreement))
+    (else (error "Invalid channel"))))
+
+(def (listen-for-agreement/libp2p libp2p-client)
+  (let* ((self (libp2p-identify libp2p-client)))
+    (for (p (peer-info->string* self))
+      (displayln "I am " p))
+    (displayln "Listening for incoming connections")
+    ;; TODO: use parameterize instead?
+    (def ret (make-parameter #f))
+    (libp2p-listen libp2p-client [chat-proto] (chat-handler ret))
+    ;; TODO close the connection
```

Comment:
Not too sure how to approach this and cleanly close the connection.
I am aware TCP will use handshakes to close the conn,
and we do actually use TCP as the transport layer protocol in the multiaddress.
But using the library function seems to cut the connection immediately, perhaps terminating it at the IP layer?
Leaving this as a TODO for now as it will take more investigating.
Have tracked this in the parent issue as well.